### ISSUE TITLE
[BUG] UtilityClass nested in an enum does not work after upgrading to Java 21

### ISSUE THREAD

**Issue by lbenedetto** [projectlombok/lombok#3679]
**Describe the bug**

After upgrading my project to Java 21, I encountered a compile error. It was difficult to find because it caused thousands of other unrelated lombok compile errors with stuff like constructors and getters not being generated. However, once I removed the single offending annotation, the project compiles without error.

**To Reproduce**

```java
@AllArgsConstructor
@Getter
public enum ExampleEnum {

  THING_ONE(ExampleEnumName.THING_ONE),
  ;

  private final String providerName;

  @UtilityClass
  public class ExampleEnumName {
    public static final String THING_ONE = "THING_ONE";
  }
}
```
On its own, this seems to work. However, when I then reference it like this:
```java
@Service(ExampleEnum.ExampleEnumName.THING_ONE)
@RequiredArgsConstructor
public class ExampleService implements ExampleServiceProvider {
}
```
I get this error:
```
/Users/lars/Work/example/UseCaseImplementations/src/main/java/ExampleService.java:4: error: cannot access ExampleEnumName
@Service(ExampleEnum.ExampleEnumName.THING_ONE)
                    ^
  bad class file: /Users/lars/Work/example/Domain/build/classes/java/main/ExampleEnum$ExampleEnumName.class
    unexpected constant pool tag: 12 at 19
    Please remove or make sure it appears in the correct subdirectory of the classpath.
```
Notice that the two classes are in different Gradle Modules (UseCaseImplementations and Domain) where UseCaseImplmentations build.gradle has
```gradle
dependencies {
    api project(':Domain')
```

**Version info (please complete the following information):**
 - Lombok version 1.18.32
 - Platform: IntelliJ with Gradle 8.7, Java 21 (21.0.3-amzn)

**Additional context**
I've worked around the issue by just removing the Utility class annotation.

**Comment by tomsit-ionos** [comment]
i ran in to this as well but in a slightly different case: 

## TLDR
**The problem occurs, if the inner class is not declared as `static`.** 
If u add the `static` modifier it went away for me.

## Long story
For me it was a nested class **inside** a record. 

It took me a good day to figure out what's going on b/c i ran into this while upgrade a keycloak extension from KC 25 -> 26.
This set me on the wrong mental path and i was guessing there were some wired upgrade problems at play.

See my stripped down reproduction repo @ https://github.com/tomsit-ionos/lombok-utilclass-bug
